src/distributed-transaction/percolator.md

```diff
@@ -0,0 +1,153 @@
+
+It means that for the key `k1`, a value `"value1"` was committed at timestamp `12`. Then there is an uncommitted version whose value is `"value2"`, and it's uncommitted because there's a lock. You will understand why it looks like this after understanding how transactions work.
+
+The data of a certain row and a certain Percolator column is named a *cell*. The above example shows different versions of data of a single cell.
+
+The remaining columns, `c:notify` and `c:ack_O`, are used for Percolator's incremental processing. After a modification, we set `c:notify` column to mark a cell to be dirty. There can be some *observers* which can do some user-specified operations when they find data of their observed columns have changed. To find whether data is changed, they continuously scanning the `notify` columns to find dirty cells. `c:ack_O` is the "acknowledgment" column of the observer `O`, which is used to prevent a row from incorrectly being notified twice. It saves the timestamp of the observer's last execution.
+
+### Writing
+
+Percolator's transactions are committed by a 2PC algorithm. Its two phases are called `Prewrite` and `Commit`.
+
+In `Prewrite` phase:
+1. We get a timestamp from the timestamp oracle, and we call the timestamp the transaction's `start_ts`.
+2. For each row involved in the transaction, put a lock to `lock` column. At the same time, put the value to the `data` column with the timestamp `start_ts`. One of these locks will be chosen as the *primary* lock, and others are *secondary* locks. Each lock saves the transaction's `start_ts`. The secondary locks, in addition, save the primary lock's location.
+    * If we found that there's already a lock or newer version than `start_ts`, the current transaction will be rolled back.
+
+And then, in `Commit` phase:
+1. Get another timestamp, namely `commit_ts`.
+2. Remove the primary lock, and at the same time write a record to `write` column with the timestamp `commit_ts`, and its value records the transaction's `start_ts`.
+    * If the lock is missing, the commit should fail.
+3. For all secondaries, do the same thing.
+
+Once the step 2 (committing the primary) is done, we say the whole transaction is done. It doesn't matter if the process of committing the secondaries unexpectedly exited. When we meet a lock of other transactions, we can see whether its primary has been committed to determine whether the whole transaction has been committed.
+
+Let's see the example in the paper of Percolator. Assume we are writing two rows in a single transaction. At first, the data looks like this:
+
+| key | bal:data    | bal:lock  | bal:write       |
+|-----|-------------|-----------|-----------------|
+| Bob | 6:<br>5:$10 | 6:<br/>5: | 6:data@5<br/>5: |
+| Joe | 6:<br/>5:$2 | 6:<br/>5: | 6:data@5<br/>5: |
+
+This table shows Bob and Joe's balance. Now Bob wants to transfer his $7 to Joe's account. First, we do `Prewrite`:
```

Comment:
```suggestion
This table shows Bob and Joe's balance. Now Bob wants to transfer his $7 to Joe's account. The first step is`Prewrite`:
```

Reply:
I think we should pay attention to using a consistent and right person in our writing. Or else it would be somewhat confusing.

Reply:
Sorry for that 😢